### PR TITLE
feat: expose GitHub polling interval setting

### PR DESCRIPTION
**Pull request by lsm**
Adds a persisted global GitHub polling interval setting (default 120s), exposes it in General settings, and removes the old GITHUB_POLLING_INTERVAL config path.

Polling cadence is now read from settings at runtime, so updates apply without restarting and 0 disables polling.

Verified:
- cd packages/daemon && bun test tests/unit/4-space-storage/storage/settings-repository.test.ts tests/unit/2-handlers/github/github-service-polling.test.ts
- cd packages/daemon && NEOKAI_USE_DEV_PROXY=1 bun test tests/online/features/github-poll-job.test.ts
- bun run check

### CODE REVIEW

**Review by lsm — COMMENTED** [review]
## 🤖 Review by claude-4-5-sonnet (Anthropic)

> **Model:** claude-4-5-sonnet | **Client:** NeoKai | **Provider:** Anthropic

**Recommendation: REQUEST_CHANGES**

The implementation is solid overall: the shared settings shape is extended, the repository backfills legacy rows with the 120s default, the daemon reads the interval live from settings, and the UI exposes the field. Unit and online tests pass, and `bun run check` is clean (pre-existing warnings only).

### Findings

**P1 — GitHub service enable condition ignores the persisted polling setting**
`packages/daemon/src/app.ts` now enables `GitHubService` whenever Anthropic auth is present:

```ts
const shouldEnableGitHub = config.githubWebhookSecret || hasAnthropicAuth;
```

This starts the service even when the user has set the polling interval to `0` and has no webhook secret, which is unnecessary and changes the previous env-var behavior. Because the default is now `120`, the equivalent persisted check should gate startup:

```ts
const shouldEnableGitHub = config.githubWebhookSecret || getGitHubPollingIntervalSeconds() > 0;
```

This keeps the default path enabled, respects a value of `0`, and still requires `hasAnthropicAuth` via the outer guard.

**P2 — UI saves the polling interval on every keystroke**
`packages/web/src/components/settings/GeneralSettings.tsx` fires `updateGlobalSettings` inside `onInput`, so typing e.g. "300" triggers three RPCs + daemon refreshes and briefly disables the input via `isUpdating`. Consider debouncing or saving on `onBlur` for numeric inputs.

Please address the P1; the P2 is a UX suggestion.

File: packages/daemon/src/app.ts [hunk 1]
```diff
@@ -595,9 +600,7 @@ export async function createDaemonApp(options: CreateDaemonAppOptions): Promise<
     startupPhase('github service');
     // Initialize GitHub service if configured
     let gitHubService: GitHubService | null = null;
-    const shouldEnableGitHub =
-      config.githubWebhookSecret ||
-      (config.githubPollingInterval && config.githubPollingInterval > 0);
+    const shouldEnableGitHub = config.githubWebhookSecret || hasAnthropicAuth;
```
Comment (lsm) [inline packages/daemon/src/app.ts]: P1: This ignores the persisted polling setting. The service starts whenever Anthropic auth is present, even if the user sets the interval to `0` and has no webhook secret. Use `config.githubWebhookSecret || getGitHubPollingIntervalSeconds() > 0` so the setting gates startup while keeping the default `120` path enabled.

File: packages/web/src/components/settings/GeneralSettings.tsx [hunk 2]
```diff
@@ -117,6 +121,34 @@ export function GeneralSettings() {
     }
   };
 
+  const handleGitHubPollingIntervalChange = async (value: string) => {
```
Comment (lsm) [inline packages/web/src/components/settings/GeneralSettings.tsx]: P2 (UX): Saving on every `onInput` keystroke triggers an RPC + daemon refresh for each digit typed (e.g. "300" = 3 updates) and briefly disables the field. Consider debouncing or saving on `onBlur` for numeric inputs.

File: packages/daemon/src/app.ts (hunk 1 above)
Comment (lsm) [reply packages/daemon/src/app.ts]: Fixed: startup gating is back to `config.githubWebhookSecret || getGitHubPollingIntervalSeconds() > 0`, so setting the persisted interval to 0 with no webhook secret prevents GitHubService startup.

File: packages/web/src/components/settings/GeneralSettings.tsx (hunk 2 above)
Comment (lsm) [reply packages/web/src/components/settings/GeneralSettings.tsx]: Fixed: the numeric input now only updates local state while typing and persists on blur after validation, avoiding per-keystroke RPCs and field disabling/flicker.

**Review by lsm — COMMENTED** [review]
## 🤖 Review by claude-4-5-sonnet (Anthropic)

> **Model:** claude-4-5-sonnet | **Client:** NeoKai | **Provider:** Anthropic

**Recommendation: APPROVE**

All previous review feedback has been addressed:

- `packages/daemon/src/app.ts` now gates `GitHubService` startup on `config.githubWebhookSecret || getGitHubPollingIntervalSeconds() > 0`, respecting a persisted interval of `0` while keeping the default `120` path enabled.
- `packages/web/src/components/settings/GeneralSettings.tsx` now persists the polling interval on blur after validation, avoiding per-keystroke RPCs.

Verification:
- Re-ran daemon unit tests for settings + GitHub polling + event extension: 135 pass
- Re-ran `NEOKAI_USE_DEV_PROXY=1 bun test tests/online/features/github-poll-job.test.ts`: 6 pass
- Re-ran `bun run check`: clean (only pre-existing oxlint warnings in workflow-migration.ts)
- Confirmed all PR review threads are resolved

No remaining findings.

**Review by lsm — COMMENTED** [review]
(This review's body was lost: the reviewer invoked `gh api -f body=@-`, but gh api does not read stdin via `@-` — it posts the literal string "@-" and discards the heredoc. The reviewer system contract has been corrected to use a quoted-heredoc command-substitution form.)

File: packages/daemon/src/app.ts [hunk 3]
```diff
@@ -680,11 +682,37 @@ export async function createDaemonApp(options: CreateDaemonAppOptions): Promise<
...
+      | GitHubEventExtension
+      | undefined;
+    internalEventBus.subscribe(
+      'settings.updated',
+      (event) => {
+        if (event.namespaceId !== 'global' || event.settings.githubPollingInterval === undefined)
+          return;
+        gitHubService?.refreshPolling();
+        void (async () => {
+          if (getGitHubPollingIntervalSeconds() > 0) {
```
Comment (lsm) [inline packages/daemon/src/app.ts]: P2: Add an `else` branch that sets `capabilities.polling: false` when the interval is `0`. Otherwise the Space GitHub UI still shows polling as available while no timers are running, leaving watched repos configured as polling-enabled but never polled.

File: packages/daemon/src/lib/github/github-service.ts [hunk 4]
```diff
@@ -167,39 +185,44 @@ export class GitHubService {
...
-        if (existing.length === 0) {
-          this.jobQueue.enqueue({ queue: GITHUB_POLL, payload: {}, runAt: Date.now() });
-          log.info('Enqueued initial github.poll job');
-        }
+      const existing = this.jobQueue.listJobs({
+        queue: GITHUB_POLL,
+        status: ['pending', 'processing'],
+        limit: 1,
+      });
+      if (existing.length === 0) {
```
Comment (lsm) [inline packages/daemon/src/lib/github/github-service.ts:219]: P2: An existing pending `github.poll` job keeps its old `runAt` here. When the interval changes (especially lowered), the new cadence does not take effect until that stale job fires. Update or replace the pending job so interval changes apply promptly.

**Review by lsm — COMMENTED** [review]
(This review's body was lost: the reviewer invoked `gh api -f body=@-`, but gh api does not read stdin via `@-` — it posts the literal string "@-" and discards the heredoc. The reviewer system contract has been corrected to use a quoted-heredoc command-substitution form.)

**Review by lsm — COMMENTED** [review]
(This review's body was lost: the reviewer invoked `gh api -f body=@-`, but gh api does not read stdin via `@-` — it posts the literal string "@-" and discards the heredoc. The reviewer system contract has been corrected to use a quoted-heredoc command-substitution form.)

File: packages/daemon/src/app.ts [hunk 5]
```diff
@@ -680,11 +682,34 @@ export async function createDaemonApp(options: CreateDaemonAppOptions): Promise<
...
     );
 
+    const githubEventExtension = extensionManager.getExtension('github') as
+      | GitHubEventExtension
+      | undefined;
+    internalEventBus.subscribe(
+      'settings.updated',
+      (event) => {
+        if (event.namespaceId !== 'global') return;
+        gitHubService?.refreshPolling({ reschedulePending: true });
```
Comment (lsm) [inline packages/daemon/src/app.ts:713]: P2: Track the last effective `githubPollingInterval` and only reschedule/delete pending jobs when the interval actually changes. Otherwise every unrelated global settings edit (Default Model, Auto-scroll, etc.) replaces the pending `github.poll` job with an immediate one and increases poll frequency.

**Review by lsm — COMMENTED** [review]
(This review's body was lost: the reviewer invoked `gh api -f body=@-`, but gh api does not read stdin via `@-` — it posts the literal string "@-" and discards the heredoc. The reviewer system contract has been corrected to use a quoted-heredoc command-substitution form.)

**Review by lsm — COMMENTED** [review]
(This review's body was lost: the reviewer invoked `gh api -f body=@-`, but gh api does not read stdin via `@-` — it posts the literal string "@-" and discards the heredoc. The reviewer system contract has been corrected to use a quoted-heredoc command-substitution form.)

File: packages/daemon/src/app.ts [hunk 6]
```diff
@@ -595,9 +600,7 @@ export async function createDaemonApp(options: CreateDaemonAppOptions): Promise<
     startupPhase('github service');
     // Initialize GitHub service if configured
     let gitHubService: GitHubService | null = null;
-    const shouldEnableGitHub =
-      config.githubWebhookSecret ||
-      (config.githubPollingInterval && config.githubPollingInterval > 0);
+    const shouldEnableGitHub = config.githubWebhookSecret || getGitHubPollingIntervalSeconds() > 0;
```
Comment (lsm) [inline packages/daemon/src/app.ts:620]: P2: Construct GitHubService whenever Anthropic auth is available so the service can react live when the persisted interval changes from 0 to positive. Currently it remains null in that scenario and the settings subscriber cannot start job-queue polling without a restart.

**Review by lsm — COMMENTED** [review]
(This review's body was lost: the reviewer invoked `gh api -f body=@-`, but gh api does not read stdin via `@-` — it posts the literal string "@-" and discards the heredoc. The reviewer system contract has been corrected to use a quoted-heredoc command-substitution form.)

File: packages/daemon/src/app.ts [hunk 7]
```diff
@@ -665,9 +673,7 @@ export async function createDaemonApp(options: CreateDaemonAppOptions): Promise<
       },
     };
     const extensionManager = new ExternalEventExtensionManager();
-    const githubPollingEnabled = !!(
-      config.githubPollingInterval && config.githubPollingInterval > 0
-    );
+    const githubPollingEnabled = getGitHubPollingIntervalSeconds() > 0;
```
Comment (lsm) [inline packages/daemon/src/app.ts:693]: P2: This block only sets `capabilities.polling: true` when the global interval is positive. If the daemon restarts with the interval set to `0` while the persisted GitHub extension config still has `capabilities.polling: true` (e.g. from a prior run or an unclean shutdown before the settings subscriber could write `false`), the stale `true` value remains in the DB. The Space GitHub UI derives polling availability from that flag, so it can continue to show polling as available even though every poll path is gated by `getPollIntervalMs() <= 0` and the `setPollingEnabled`/`watchRepo` RPCs reject enablement.

Add an `else` branch that writes `polling: false` when `githubPollingEnabled` is false, or factor the capability update into a helper that runs both at startup and on settings changes.

**Review by lsm — COMMENTED** [review]
## 🤖 Review by Claude (Anthropic)

> **Model:** Claude | **Client:** NeoKai | **Provider:** Anthropic

Recommendation: REQUEST_CHANGES

The latest iteration correctly guards space/repo polling enablement when the global interval is `0`, and the job-queue handler now recomputes the interval after an in-flight poll. Targeted unit tests pass and quality checks are green.

However, one previously raised edge case is still incomplete: the startup capability initialization in `packages/daemon/src/app.ts` only writes `capabilities.polling: true` when the interval is positive, but does not write `false` when the interval is `0`. If the persisted extension config already has `polling: true` (prior run, unclean shutdown, etc.), a restart with interval `0` leaves the flag stale, so the Space UI can still present polling as available.

Please add the startup `else` branch (or share the capability-update logic with the `settings.updated` subscriber) and add a regression test, then re-verify the targeted tests and `bun run check`.

Verification run:
- `cd packages/daemon && bun test tests/unit/2-handlers/github/github-service-polling.test.ts tests/unit/2-handlers/github/github-event-extension.test.ts` — 102 pass
- `bun run check` — pass (pre-existing lint warnings in `workflow-migration.ts` only)

File: packages/daemon/src/lib/external-events/github/github-event-extension.ts [hunk 8]
```diff
@@ -262,6 +263,9 @@ export class GitHubEventExtension implements HttpExternalEventExtension, RpcExte
       if (!params.spaceId || !params.owner || !params.repo) {
         throw new Error('spaceId, owner and repo are required');
       }
+      if (params.pollingEnabled) {
```
Comment (lsm) [inline packages/daemon/src/lib/external-events/github/github-event-extension.ts]: P2: This guard rejects any `watchRepo` call that carries `pollingEnabled: true`, including updates that merely preserve an already-polling row (e.g. toggling webhook/enabled state). Compare against the existing row and only reject a false-to-true transition so unrelated updates to existing polling repos are not blocked when the global interval is `0`.

File: packages/daemon/src/lib/external-events/github/github-event-extension.ts [hunk 9]
```diff
@@ -651,8 +658,20 @@ export class GitHubEventExtension implements HttpExternalEventExtension, RpcExte
...
   }
 
+  async refreshPollingInterval(): Promise<void> {
+    if (this.stopped) return;
+    if (!(await this.isPollingGloballyEnabled()) || this.getPollIntervalMs() <= 0) {
+      if (this.pollTimer) clearTimeout(this.pollTimer);
+      this.pollTimer = null;
+      return;
+    }
+    this.scheduleNextPoll();
```
Comment (lsm) [inline packages/daemon/src/lib/external-events/github/github-event-extension.ts]: P2: `refreshPollingInterval()` schedules the next poll using the configured interval without checking `rateLimitedUntil`. If the user edits the interval while GitHub has asked us to back off until a time later than that interval, this cancels the rate-limit deferral and risks polling during the cooldown. Compute the delay from `Math.max(this.getPollIntervalMs(), this.rateLimitedUntil - Date.now())` (with the same flooring logic as `runPollCycle`).

**Review by lsm — COMMENTED** [review]
## 🤖 Review by Claude (Anthropic)

> **Model:** Claude | **Client:** NeoKai | **Provider:** Anthropic

Recommendation: APPROVE

All four edge-case fixes are in place and verified:

- Startup now syncs `capabilities.polling` bidirectionally via `syncGitHubPollingCapability`, clearing stale `true` when the global interval is `0`.
- `space.github.enable` refuses to re-arm the polling capability while the global interval is `0`.
- `watchRepo` only rejects false → true polling transitions, allowing unrelated updates that preserve an existing polling flag.
- `refreshPollingInterval()` uses `getNextPollDelayMs()`, preserving active `rateLimitedUntil` deferrals.

Regression tests were added for each scenario. Verification:
- `cd packages/daemon && bun test tests/unit/2-handlers/github/github-service-polling.test.ts tests/unit/2-handlers/github/github-event-extension.test.ts` — 106 pass
- `bun run check` — pass (pre-existing lint warnings in `workflow-migration.ts` only)
- All 20 PR review threads resolved.

**Review by lsm — COMMENTED** [review]
## 🤖 Review by Claude (Anthropic)

> **Model:** Claude | **Client:** NeoKai | **Provider:** Anthropic

Recommendation: REQUEST_CHANGES

Withdrawing the prior approval: a fresh bot comment correctly identifies that the `space.github.enable` interval guard is placed after the DB mutation. When the global interval is `0` and a disabled space has polling rows, `setRepoEnabled(..., true)` already flips the space/repository `enabled` flags before `assertPollingIntervalEnabled()` throws, leaving the DB partially re-enabled and inconsistent.

Please move the interval check before `this.repo.setRepoEnabled(params.spaceId, true)` in `space.github.enable`, and add/update the regression test to assert that no DB mutation occurs when the RPC rejects.

After fixing, re-run the targeted GitHub daemon tests and `bun run check`.

Referenced comment: https://github.com/lsm/neokai/pull/2157#discussion_r3461808427

**Review by lsm — COMMENTED** [review]
## 🤖 Review by Claude (Anthropic)

> **Model:** Claude | **Client:** NeoKai | **Provider:** Anthropic

Recommendation: APPROVE

The mutation-order finding is fixed: `space.github.enable` now checks for polling-configured repos and asserts the global interval is positive before calling `setRepoEnabled(..., true)`, so interval `0` rejects without mutating the DB. The regression test asserts the watched repo remains disabled after the rejected RPC.

All previously raised edge cases are addressed:
- Startup syncs `capabilities.polling` both ways via `syncGitHubPollingCapability`.
- `space.github.enable` guards interval `0` and fails fast without mutation.
- `watchRepo` only rejects false → true polling transitions.
- `refreshPollingInterval()` preserves active rate-limit deferrals.

Verification:
- `cd packages/daemon && bun test tests/unit/2-handlers/github/github-service-polling.test.ts tests/unit/2-handlers/github/github-event-extension.test.ts` — 111 pass
- `bun run check` — pass (pre-existing lint warnings in `workflow-migration.ts` only)
- All 21 PR review threads resolved

**Review by lsm — COMMENTED** [review]
## 🤖 Review by Claude (Anthropic)

> **Model:** Claude | **Client:** NeoKai | **Provider:** Anthropic

Recommendation: REQUEST_CHANGES

Withdrawing the approval again: a fresh bot comment correctly identifies that the `space.github.enable` guard is too broad. Rejecting the entire space re-enable when any row has `pollingEnabled: true` and the global interval is `0` also blocks webhook-only recovery for spaces that contain webhook-configured repos alongside polling-flagged rows. The global interval already prevents timers from running, so the user should not have to remove rows or re-enable polling globally just to restore webhook delivery for the space.

Please change `space.github.enable` so it does not reject the RPC when interval is `0`; instead, conditionally re-arm the polling capability/timer only when `getPollIntervalMs() > 0`. Update the regression test to assert that the space/repos are enabled and webhook-configured repos are functional, while no polling capability/timer is re-armed.

Referenced comment: https://github.com/lsm/neokai/pull/2157#discussion_r3461954212

**Review by lsm — COMMENTED** [review]
## 🤖 Review by Claude (Anthropic)

> **Model:** Claude | **Client:** NeoKai | **Provider:** Anthropic

Recommendation: APPROVE

Both latest edge cases are fixed:

- `space.github.enable` now re-enables the space when interval is `0` while skipping only the polling capability/timer re-arm, preserving webhook delivery recovery.
- `refreshPollingInterval()` returns early when `activePollCycle` is in flight, letting the active cycle self-schedule after completion with the live interval and avoiding overlapping polls.

All previously raised issues remain addressed (startup capability sync, false-to-true polling transition guard, rate-limit deferral preservation, in-flight poll interval recomputation, pending job replacement).

Verification:
- `cd packages/daemon && bun test tests/unit/2-handlers/github/github-service-polling.test.ts tests/unit/2-handlers/github/github-event-extension.test.ts` — 112 pass
- `bun run check` — pass (pre-existing lint warnings in `workflow-migration.ts` only)
- All 23 PR review threads resolved
